File: RLBotPack/Rocketnoodles/src/world/components/game.py
from world.components import WorldComponent
from dataclasses import dataclass
from rlbot.utils.structures.game_data_struct import GameInfo


@dataclass
class Game(WorldComponent):
    """"Game Component for the world model.

    :param info: Metadata about a single match.
    :type GameInfo:
    """

    game_speed: float
    game_time_remaining: float
    is_kickoff_pause: bool
    is_match_ended: bool
    is_overtime: bool
    is_round_active: bool
    is_unlimited_time: bool
    seconds_elapsed: float
    world_gravity_z: float

    def __init__(self, info: GameInfo):
        self.update(info)

    def update(self, info: GameInfo):
        """Update function for the Game class. Updates the internal state of this component to the current step.

        :param info: Metadata about a single match.
        :type info: GameInfo
        """

        self.game_speed = info.game_speed
        self.game_time_remaining = info.game_time_remaining
        self.is_kickoff_pause = info.is_kickoff_pause
        self.is_match_ended = info.is_match_ended
        self.is_overtime = info.is_overtime
        self.is_round_active = info.is_round_active
        self.is_unlimited_time = info.is_unlimited_time
        self.seconds_elapsed = info.seconds_elapsed
        self.world_gravity_z = info.world_gravity_z

File: RLBotPack/PhoenixCS/RedUtils/Objects/Game.cs
using RedUtils.Math;
using rlbot.flat;
using System;

namespace RedUtils
{
	/// <summary>Contains basic info on the current match</summary>
	public static class Game
	{
		/// <summary>The scores for the blue team, and orange team (in that order)</summary>
		public static int[] Scores { get; private set; }
		/// <summary>The blue team's score</summary>
		public static int BlueScore => Scores[1];
		/// <summary>The orange team's score</summary>
		public static int OrangeScore => Scores[0];

		/// <summary>How much time has passed since the game has began</summary>
		public static float Time { get; private set; }
		/// <summary>How much time is remaining before the game ends</summary>
		public static float TimeRemaining { get; private set; }
		/// <summary>The speed at which game plays</summary>
		public static float GameSpeed { get; private set; }

		/// <summary>Whether or not the game is configured with unlimited time</summary>
		public static bool IsUnlimitedTime { get; private set; }
		/// <summary>Whether or not the game is currently in overtime</summary>
		public static bool IsOvertime { get; private set; }
		/// <summary>Whether or not a round is currently active (meaning we aren't in a replay)</summary>
		public static bool IsRoundActive { get; private set; }
		/// <summary>Whether or not the game/timer has paused for kickoff</summary>
		public static bool IsKickoffPause { get; private set; }
		/// <summary>Whether or not the game has concluded</summary>
		public static bool IsMatchEnded { get; private set; }

		/// <summary>The acceleration caused by gravity</summary>
		public static Vec3 Gravity { get; private set; }

		static Game()
		{
			Scores = new int[2] { 0, 0 };;

			Time = 0;
			TimeRemaining = 300;
			GameSpeed = 1;

			IsUnlimitedTime = false;
			IsOvertime = false;
			IsRoundActive = false;
			IsKickoffPause = false;
			IsMatchEnded = false;

			Gravity = new Vec3(0, 0, -650);
		}

		/// <summary>Updates info about the game using data from the packet</summary>
		public static void Update(GameTickPacket packet)
		{
			Scores = new int[2] { packet.Teams(0).Value.Score, packet.Teams(1).Value.Score };

			Time = packet.GameInfo.Value.SecondsElapsed;
			TimeRemaining = packet.GameInfo.Value.GameTimeRemaining;
			GameSpeed = packet.GameInfo.Value.GameSpeed;

			IsUnlimitedTime = packet.GameInfo.Value.IsUnlimitedTime;
			IsOvertime = packet.GameInfo.Value.IsOvertime;
			IsRoundActive = packet.GameInfo.Value.IsRoundActive;
			IsKickoffPause = packet.GameInfo.Value.IsKickoffPause;
			IsMatchEnded = packet.GameInfo.Value.IsMatchEnded;

			Gravity = new Vec3(0, 0, packet.GameInfo.Value.WorldGravityZ);
		}
	}
}
